src/sentry/runner/commands/backup.py

```diff
@@ -59,7 +60,90 @@ def pretty(self) -> str:
         return "\n".join(f.pretty() for f in self.findings)
 
 
-def validate(expect: JSONData, actual: JSONData) -> ComparatorFindings:
+class JSONScrubbingComparator(ABC):
+    """An abstract class that compares and then scrubs some set of fields that, by a more nuanced
+    definition that mere strict byte-for-byte equality, are expected to maintain some relation on
+    otherwise equivalent JSON instances of the same model.
+
+    Each class inheriting from `JSONScrubbingComparator` should override the abstract `compare`
+    method with its own comparison logic. The `scrub` method is universal (it merely moves the
+    compared fields from the `fields` dictionary to the non-diffed `scrubbed` dictionary).
+
+    If multiple comparators are used sequentially on a single model (see the `SCRUBBING_COMPARATORS`
+    dict below for specific mappings), all of the `compare(...)` methods are called before any of
+    the `scrub(...)` methods are. This ensures that comparators that touch the same fields do not
+    have their inputs mangled by one another."""
+
+    def __init__(self, fields: list[str]):
+        self.fields = fields
+
+    def check(self, side: str, data: JSONData) -> None:
+        """Ensure that we have received valid JSON data at runtime."""
+
+        if "model" not in data or not isinstance(data["model"], str):
+            raise RuntimeError(f"The {side} input must have a `model` string assigned to it.")
+        if "pk" not in data or not isinstance(data["pk"], int):
+            raise RuntimeError(f"The {side} input must have a numerical `pk` entry.")
+        if "fields" not in data or not isinstance(data["fields"], dict):
+            raise RuntimeError(f"The {side} input must have a `fields` dictionary.")
+
+    @abstractmethod
+    def compare(self, on: InstanceID, left: JSONData, right: JSONData) -> ComparatorFinding | None:
+        """An abstract method signature, to be implemented by inhering classes with their own comparison logic."""
+
+        pass
+
+    def scrub(self, on: InstanceID, left: JSONData, right: JSONData) -> None:
+        """Removes all of the fields compared by this comparator from the `fields` dict, so that the remaining fields may be compared for equality."""
+
+        self.check("left", left)
+        self.check("right", right)
+        if "scrubbed" not in left:
```

Comment:
What is `left` here? Do we risk (at least theoretically) that there's already a `scrubbed` key?

Reply:
I mean, it's a `JSONData`, but that's pretty general. Looking at `check` above, I'm like 80% sure that we're not at risk here (i.e., we're not sharing a namespace with the model fields themselves, i.e., something defined somewhere else that we don't control).

Reply:
`left` and `right` are the two JSON model entries being compared - I've expanded the comment to illuminate this.

In terms of keys already being scrubbed: I think the risk is low, because the actual comparison code near the bottom of the `validate()` function takes care to always run all of the `compare()` methods before any of the `scrub()` methods for exactly this reason, to ensure that each comparator receives data that is unmodified by its predecessor. This means that comparators should be able to be safely listed in `DEFAULT_COMPARATORS` entries in any order without changing the output. I've added a comment to that method below to ensure that future maintainers of this code don't mess up this invariant.